### PR TITLE
feat(platform): re-attempt platform automerge on github and gitlab 

### PR DESCRIPTION
**Pull request by straub**
## Changes

This change will cause Renovate to re-enable GitHub's automerge / Merge Queue feature after rebasing the PR, to more seamlessly recover when PRs encounter merge conflicts in the queue and are removed by `github-merge-queue[bot]`.

This is a naive first pass at the feature, and I'm submitting this PR as a work-in-progress to get feedback. I suspect, among other potential feedback, that my testing needs to be improved and that it's possible the new feature should be opt-in via a new config flag (not yet added).

Thanks for your time!

## Context

https://github.com/renovatebot/renovate/discussions/26523

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by HonkingGoose** [comment]
> I'm unsure whether this is needed. I have not included a doc update in this PR yet.

The new behavior should be explained in [Renovate's `platformAutomerge` docs](https://docs.renovatebot.com/configuration-options/#platformautomerge). Currently those docs say:

> Renovate tries platform-native automerge only when it initially creates the PR. Any PR that is being updated will be automerged with the Renovate-based automerge.

But your PR makes it so that Renovate enables the platform native automerge (GitHub Merge queue in this case) again after a merge conflict, I think. So that should be explained.

File: lib/modules/platform/bitbucket-server/index.ts
Comment (HonkingGoose) [inline lib/modules/platform/bitbucket-server/index.ts:377]: [Renovate's `platformAutomerge` docs](https://docs.renovatebot.com/configuration-options/#platformautomerge) should be updated, once the code is done.

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: lib/modules/platform/bitbucket-server/index.ts
Comment (rarkins) [inline lib/modules/platform/bitbucket-server/index.ts:378]: Is this renaming necessary? I think either keep it as-is, or change the type itself

**Review by rarkins — COMMENTED** [review]
I think that the refreshPr for bitbucket is such a different use - and called at different times where re-enabling automerge is not relevant - that it's probably not a good idea to combine it

File: lib/modules/platform/bitbucket-server/index.ts
Comment (straub) [reply lib/modules/platform/bitbucket-server/index.ts:378]: This was me trying to keep consistency with `updatePr()`, I'll iron it out on the next pass 👍

File: lib/modules/platform/bitbucket-server/index.ts
Comment (straub) [reply lib/modules/platform/bitbucket-server/index.ts]: I'm struggling a bit to come up with a way to phrase this that's up to the high quality bar set by the rest of the docs, but I made an attempt in the last push!

**Comment by straub** [comment]
I've refactored this implementation to avoid using/changing `refreshPr`. Now using a new method called `reattemptPlatformAutomerge()`

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
Needs more work after applying the suggestion.

File: lib/modules/platform/github/index.ts
Comment (viceice) [inline lib/modules/platform/github/index.ts:1746]: won't this merge a PR which doesn't use platform automerge? we don't want that on this stage.

Comment (straub) [reply lib/modules/platform/github/index.ts:1746]: My understanding is: no. `tryPrAutomerge` contains all the checks for whether platform automerge should be used, so it'll do nothing as appropriate

Comment (rarkins) [reply lib/modules/platform/github/index.ts:1746]: The problem might be that you're doing an extra API call every rebase even if no automerge platform automerge is enabled

Comment (rarkins) [reply lib/modules/platform/github/index.ts:1746]: I think this check could be done up in the application logic, e.g. if config.automerge and config.platformAutomerge and platform.reattemptPlatformAutomerge then.....

Comment (straub) [reply lib/modules/platform/github/index.ts:1746]: Ah, I thought the API call was cached at this point so that it wouldn't fire again.

I've added the checks from `tryPrAutomerge` up higher in the last push. I didn't add a check for `config.automerge` because I saw `createPr` doesn't do that, and I wanted this feature to have the same behavior, but please let me know if you'd like that included!

https://github.com/renovatebot/renovate/blob/main/lib/modules/platform/github/index.ts#L1686

File: lib/modules/platform/bitbucket-server/index.ts
Comment (straub) [reply lib/modules/platform/bitbucket-server/index.ts]: I just noticed that the GitLab platform calls its `tryPrAutomerge` during both `createPr` and `updatePr`. Perhaps the sentence "Renovate tries platform-native automerge only when it initially creates the PR." in the doc was already incomplete or out-of-date?

https://github.com/renovatebot/renovate/blob/0b3b56adb555ae993a8f89ea721c5dfa36c4b259/lib/modules/platform/gitlab/index.ts#L818

File: lib/modules/platform/github/index.ts
Comment (straub) [reply lib/modules/platform/github/index.ts:1746]: Looking one level deeper, I see `getPlatformPrOptions` accounts for `config.automerge` (among other things) when determining its `usePlatformAutomerge` value, so this should be fully covered.

https://github.com/renovatebot/renovate/blob/0b3b56adb555ae993a8f89ea721c5dfa36c4b259/lib/workers/repository/update/pr/index.ts#L50-L54

File: lib/workers/repository/update/branch/index.ts
Comment (rarkins) [inline lib/workers/repository/update/branch/index.ts:578]: Are you sure that this is visible at the worker level it's used in here? I thought it might only be visible in platform/github/index.ts

File: lib/workers/repository/update/branch/index.ts
Comment (straub) [inline lib/workers/repository/update/branch/index.ts:577]: It's mystifying at the moment me how `platformOptions` could be undefined right here, but `lib/workers/repository/update/branch/index.spec.ts` fails if I don't use optional chaining in this spot...

File: lib/workers/repository/update/branch/index.ts
Comment (straub) [reply lib/workers/repository/update/branch/index.ts:578]: I'm not terribly sure; I'll admit I've been leaning on Typescript a lot for this change. I'll try to figure that out!

If not, I think checking only `usePlatformAutomerge` would still broadly accomplish the goal of "avoid an extra API call when automerge isn't enabled"?

File: lib/workers/repository/update/branch/index.ts
Comment (straub) [reply lib/workers/repository/update/branch/index.ts:577]: Removing the upper check on `config.autoMergeAllowed` doesn't fix coverage. Going into a block of meetings, but will come back to this later

Comment (straub) [reply lib/workers/repository/update/branch/index.ts:577]: Experimentally verified `platformOptions === undefined` here and `getPlatformPrOptions()` doesn't seem to be called (a throw inside it doesn't change the outcome of `lib/workers/repository/update/branch/index.spec.ts` tests). I don't use Typescript at my day job, so maybe I'm making some rookie mistake?

Comment (straub) [reply lib/workers/repository/update/branch/index.ts:577]: Ahhh, Jest mocks! Another thing I don't use at my day job 😅 So when `lib/workers/repository/update/branch/index.spec.ts` runs, the `lib/workers/repository/update/pr` module is mocked away, including `getPlatformPrOptions`.

This is telling me I do have a very real gap in my test coverage here...

File: lib/workers/repository/update/branch/index.ts
Comment (straub) [reply lib/workers/repository/update/branch/index.ts]: Following the flow from `renovateRepository` > `initRepo` > `updateRepo` > `update` > `writeUpdates` > `processBranch`, I wasn't quite able to understand if `config.autoMergeAllowed` would be visible here or not. Still, it does seem like a check that shouldn't happen at this level, so I removed it in my latest push

File: lib/modules/platform/bitbucket-server/index.ts
Comment (straub) [reply lib/modules/platform/bitbucket-server/index.ts]: Now realizing the above, I wonder if I should rework this whole change for consistency with the gitlab behavior, and either:
1. add a call to `tryPrAutomerge` in the github platform's `updatePr`, or 
2. refactor gitlab to have and use the new `reattemptPlatformAutomerge` method instead.

File: lib/modules/platform/github/index.ts
Comment (viceice) [inline lib/modules/platform/github/index.ts:1744]: I think you should fetch the PR from cache, so we don't need this extra API call.

@zharinov do we've the `node_id` on PR cache?

File: lib/modules/platform/github/index.ts
Comment (viceice) [inline lib/modules/platform/github/index.ts:1743]:
```suggestion
    const result = getPr(prNo);
```
this will fetch from cache when available and uses rest if not on cache. so you probably don't need to change tests.

File: lib/modules/platform/github/index.ts
Comment (straub) [reply lib/modules/platform/github/index.ts]: Looks like it's not on the type, at least.

> lib/modules/platform/github/index.ts:1741:21 - error TS2339: Property 'node_id' does not exist on type 'Promise<GhPr | null>'.

I'll poke at this a bit

File: docs/usage/configuration-options.md
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:2983]:
```suggestion
On GitHub, Renovate will try to  automerge its own PR again, after the PR was rebased.
This is how Renovate fixes merge conflicts in the Merge Queue.
```

How about something like this?

File: lib/modules/platform/github/index.ts
Comment (viceice) [inline lib/modules/platform/github/index.ts:1740]:
```suggestion
    const result = await getPr(prNo);
```

File: lib/modules/platform/github/index.ts
Comment (viceice) [inline lib/modules/platform/github/index.ts:1740]:
```suggestion
    const result = (await getPr(prNo))!;
```

I think we can assume that the pr exists

File: lib/modules/platform/bitbucket-server/index.ts
Comment (straub) [reply lib/modules/platform/bitbucket-server/index.ts]: @HonkingGoose @rarkins @viceice Any thoughts on the above? I'm happy to keep moving as-is or change my approach

Comment (HonkingGoose) [reply lib/modules/platform/bitbucket-server/index.ts]: I'll let a maintainer respond. 🙂

File: docs/usage/configuration-options.md
Comment (HonkingGoose) [reply docs/usage/configuration-options.md:3008]: I don't know if we should call out Gitlab extra here.

**Review by rarkins — COMMENTED** [review]
LGTM but needs tests passing

**Comment by straub** [comment]
> LGTM but needs tests passing

Test are now passing! Please let me know if this needs anything else

**Review by rarkins — COMMENTED** [review]
Is there still any concern about feature parity with gitlab? I recall a comment being added her earlier about it

**Comment by straub** [comment]
> Is there still any concern about feature parity with gitlab? I recall a comment being added her earlier about it

I was looking for some direction on that in relation to the docs:
https://github.com/renovatebot/renovate/pull/26567#discussion_r1449013587
https://github.com/renovatebot/renovate/pull/26567#discussion_r1455744970

It's not strictly a part of this change, but I'd be happy to cover it while I'm in here

File: lib/modules/platform/bitbucket-server/index.ts
Comment (rarkins) [reply lib/modules/platform/bitbucket-server/index.ts]: I have lost track of how/why gitlab's approach is like it is. It's possible that the docs are inaccurate.

Isn't updatePr() only called if the PR title or body needs changing? i.e. a force push which does not require a change of PR title or body won't result in automerge being re-enabled?

Comment (straub) [reply lib/modules/platform/bitbucket-server/index.ts]: From [some digging just now](https://github.com/renovatebot/renovate/blob/99467af092fb621be56f515e4b7aed7d6ccaf51e/lib/workers/repository/update/pr/index.ts#L347-L351), it looks like we'll get to `updatePr` if the base branch, title, or body change, so
> i.e. a force push which does not require a change of PR title or body won't result in automerge being re-enabled?

does seem like a pretty accurate description. Still sounds inconsistent with the docs "only when it initially creates the PR" wording in GitLab's case, but I don't have personal experience with GitLab

Comment (rarkins) [reply lib/modules/platform/bitbucket-server/index.ts]: I see the GitLab retry automerge was added quite long ago by @viceice: https://github.com/renovatebot/renovate/pull/7871

@straub thanks for pointing this out and apologies that it's holding up this PR, but I think you're right that we should resolve this inconsistency.

When designing this feature, we postured that re-enabling automerge after commits was the safest/most conservative approach, because we want to avoid flip/flopping if a user tries disabling automerge and Renovate keeps turning it back on.

Meanwhile we rediscover that GitLab apparently re-enables automerge in the most common scenario where the PR title or body changes, and I guess we haven't heard complaints from users.

I think that the force push approach seems best, and it's mostly a superset of the updatePr() use cases. There are cases such as a git conflict where there will be a force push without need to update the PR content. Meanwhile are there cases where the PR needs updating without a force push where automerge needs re-enabling? I can't think of any.

File: lib/modules/platform/github/index.ts
Comment (viceice) [inline lib/modules/platform/github/index.ts:1819]: can you please add a test instead of ignoring?

File: lib/modules/platform/bitbucket-server/index.ts
Comment (straub) [reply lib/modules/platform/bitbucket-server/index.ts]: Sorry, somehow I missed this response. I'll update GitLab to use `reattemptPlatformAutomerge` 👍

**Review by viceice — DISMISSED** [review]

File: lib/modules/platform/gitlab/index.ts
Comment (straub) [inline lib/modules/platform/gitlab/index.ts:855]: @viceice, I couldn't see a way that `tryPrAutomerge` would throw for GitLab, so I removed the try/catch to resolve the coverage issue. Let me know if that's not okay!

File: lib/modules/platform/bitbucket-server/index.ts
Comment (straub) [reply lib/modules/platform/bitbucket-server/index.ts]: @rarkins, I've got both `github` and `gitlab` platforms using `reattemptPlatformAutomerge` now, and I've updated the docs to reflect that. Thanks for your patience!

File: lib/modules/platform/gitlab/index.ts
Comment (straub) [reply lib/modules/platform/gitlab/index.ts:855]: Also, raising this since I noticed, in case you all weren't already aware, it looks like Codecov has been broken since mid-February: https://app.codecov.io/github/renovatebot/renovate

Comment (viceice) [reply lib/modules/platform/gitlab/index.ts:855]: 🤔

https://github.com/renovatebot/renovate/actions/runs/8308110045/job/22738028070

it says upload successfully.

@zharinov can you check?

Comment (zharinov) [reply lib/modules/platform/gitlab/index.ts:855]: 🤦‍♂️

Comment (zharinov) [reply lib/modules/platform/gitlab/index.ts:855]: Hey @straub, good catch, our coverage is fixed now 👍

**Review by HonkingGoose — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**
